controllers/hostedcluster/controller.go

```diff
@@ -0,0 +1,134 @@
+	hyperv1beta1 "github.com/openshift/hypershift/api/v1beta1"
+)
+
+var hostedClusters = map[string]hypershiftlogforwarder.HostedCluster{}
+
+// ClusterLogForwarderTemplateReconciler reconciles a ClusterLogForwarderTemplate object
+type HostedClusterReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	log    logr.Logger
+	Mgr    ctrl.Manager
+}
+
+//+kubebuilder:rbac:groups=logging.managed.openshift.io,resources=clusterlogforwardertemplates,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=logging.managed.openshift.io,resources=clusterlogforwardertemplates/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=logging.managed.openshift.io,resources=clusterlogforwardertemplates/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *HostedClusterReconciler) Reconcile(
+	ctx context.Context,
+	req ctrl.Request,
+) (ctrl.Result, error) {
+
+	log := logr.Logger{}.WithName("hostedcluster-controller")
+
+	hostedCluster := &hyperv1beta1.HostedCluster{}
+	if err := r.Get(ctx, req.NamespacedName, hostedCluster); err != nil {
+		// Ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification).
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	_, exist := hostedClusters[hostedCluster.Name]
+
+	hcpNamespace := fmt.Sprintf("%s-%s", hostedCluster.Namespace, hostedCluster.Name)
+	hcpName := hostedCluster.Name
+
+	if !exist {
+		// check hosted cluster status, if it's new created and ready, start the reconcile
+		newReadyCluster := hostedcluster.IsReadyHostedCluster(*hostedCluster)
+		if newReadyCluster {
+			restConfig, err := hostedcluster.BuildGuestKubeConfig(r.Client, hcpNamespace, r.log)
+			if err != nil {
+				log.Error(err, "getting guest cluster kubeconfig")
+			}
+
+			hsCluster, err := cluster.New(restConfig)
+			if err != nil {
+				log.Error(err, "creating guest cluster kubeconfig")
+			}
+
+			hostedCluster := hypershiftlogforwarder.HostedCluster{
+				Cluster:      hsCluster,
+				HCPNamespace: hcpNamespace,
+			}
+			hostedClusters[hcpName] = hostedCluster
+			rhc := hypershiftlogforwarder.HyperShiftLogForwarderReconciler{
+				Client:       hostedCluster.Cluster.GetClient(),
+				Scheme:       r.Scheme,
+				MCClient:     r.Client,
+				HCPNamespace: hostedCluster.HCPNamespace,
+			}
+
+			//Adding new hosted cluster to the manager
+			r.Mgr.Add(hostedCluster.Cluster)
+			err = ctrl.NewControllerManagedBy(r.Mgr).
+				For(&v1alpha1.HyperShiftLogForwarder{}).
+				Watches(
+					source.NewKindWithCache(&v1alpha1.HyperShiftLogForwarder{}, hostedCluster.Cluster.GetCache()),
+					&handler.EnqueueRequestForObject{},
+				).
+				Complete(&rhc)
+			if err != nil {
+
+				return ctrl.Result{}, err
+			}
+
+		}
+
+	}
+
+	//if it's deleted, stop the reconcile
```

Comment:
How do we remove the hostedCluster from the map if it is deleted? Or do we just build from empty map everytime?

Reply:
Can't rebuild it from the map need to send a shutdown signal to the existing hosted cluster controller .. otherwise manager needs to restart

Reply:
Change the approach for reconciling HCP clients 
* Create a SubManager for each HostedCluster with the kubeconfig 
* Watch hlf via new manager in `go routine` thread 
* start the manager with cancel context and cache cancel function in hostedcluster 
* once the cluster is deleted, then call the manager context cancel function